hpx/util/detail/pack_traversal_impl.hpp

```diff
@@ -142,9 +143,69 @@ namespace util {
                 return Base<NewType, decltype(allocator)>(std::move(allocator));
             }
 
+            /// Returns the default iterators of the container in case
+            /// the container was passed as an l-value reference.
+            /// Otherwise move iterators of the container are returned.
+            template <typename C, typename = void>
+            class container_accessor
+            {
+                C const& container_;
+
+            public:
+                container_accessor(C const& container)
+                  : container_(container)
+                {
+                }
+
+                auto begin() -> decltype(container_.begin())
+                {
+                    return container_.begin();
+                }
+
+                auto end() -> decltype(container_.end())
+                {
+                    return container_.end();
+                }
+            };
+            template <typename C>
+            class container_accessor<C,
+                typename std::enable_if<
+                    std::is_rvalue_reference<C&&>::value>::type>
+            {
+                C&& container_;
+
+            public:
+                container_accessor(C&& container)
+                  : container_(std::move(container))
+                {
+                }
+
+                auto begin()
+                    -> decltype(std::make_move_iterator(container_.begin()))
+                {
+                    return std::make_move_iterator(container_.begin());
+                }
+
+                auto end()
+                    -> decltype(std::make_move_iterator(container_.end()))
+                {
+                    return std::make_move_iterator(container_.end());
+                }
+            };
+
+            template <typename T>
+            container_accessor<T> container_accessor_of(T&& container)
+            {
+                // Don't use any decay here
+                return container_accessor<T>(std::forward<T>(container));
+            }
+
             /// Deduces to the type the homogeneous container is containing
             template <typename Container>
-            using element_of_t = decltype(*std::declval<Container>().begin());
+            using element_of_t = typename std::conditional<
+                std::is_rvalue_reference<Container&&>::value,
+                decltype(std::move(*(std::declval<Container>().begin()))),
+                decltype(*(std::declval<Container>().begin()))>::type;
```

Comment:
I just find this construct to be difficult to parse. Could you at least add an explaining comment, please?